jina/types/document/__init__.py

```diff
@@ -0,0 +1,194 @@
+                then deserialize a :class:`DocumentProto`; ``dict`` is given then
+                parse a :class:`DocumentProto` from it; ``str`` is given, then consider
+                it as a JSON string and parse a :class:`DocumentProto` from it; finally,
+                one can also give `DocumentProto` directly, then depending on the ``copy``,
+                it builds a view or a copy from it.
+        :param copy: when ``document`` is given as a :class:`DocumentProto` object, build a
+                view (i.e. weak reference) from it or a deep copy from it.
+        :param kwargs:
+        """
+        self._document = jina_pb2.DocumentProto()
+        if isinstance(document, jina_pb2.DocumentProto):
+            if copy:
+                self._document.CopyFrom(document)
+            else:
+                self._document = document
+        elif isinstance(document, dict):
+            json_format.ParseDict(document, self._document)
+        elif isinstance(document, str):
+            json_format.Parse(document, self._document)
+        elif isinstance(document, bytes):
+            self._document.ParseFromString(document)
+
+        for k, v in kwargs.items():
+            if hasattr(self._document, k):
+                setattr(self._document, k, v)
+
+    def __getattr__(self, name: str):
+        if hasattr(_empty_doc, name):
+            return getattr(self._document, name)
+        else:
+            raise AttributeError
+
+    def update_id(self):
+        self._document.id = new_doc_id(self._document)
+
+    @property
+    def id_in_hash(self) -> int:
+        return id2hash(self._document.id)
+
+    @property
+    def id_in_bytes(self) -> bytes:
+        return id2bytes(self._document.id)
+
+    @property
+    def id(self) -> bytes:
+        return self._document.id
+
+    @id.setter
+    def id(self, value: str):
+        pass
+
+    @property
+    def blob(self) -> 'np.ndarray':
+        return NdArray(self._document.blob).value
+
+    @blob.setter
+    def blob(self, value: Union['np.ndarray', 'jina_pb2.NdArrayProto', 'NdArray']):
+        self._update_ndarray('blob', value)
+
+    @property
+    def embedding(self) -> 'np.ndarray':
+        return NdArray(self._document.embedding).value
+
+    @embedding.setter
+    def embedding(self, value: 'np.ndarray'):
+        self._update_ndarray('embedding', value)
+
+    def _update_ndarray(self, k, v):
+        if isinstance(v, jina_pb2.NdArrayProto):
+            getattr(self._document, k).CopyFrom(v)
+        elif isinstance(v, np.ndarray):
+            NdArray(getattr(self._document, k)).value = v
+        elif isinstance(v, NdArray):
+            NdArray(getattr(self._document, k)).is_sparse = v.is_sparse
+            NdArray(getattr(self._document, k)).value = v.value
+        else:
+            raise TypeError(f'{k} is in unsupported type {typename(v)}')
+
+    @property
+    def matches(self) -> Iterator['Document']:
+        for d in self._document.matches:
```

Comment:
may it be more performant to avoid having a `Generator` here? After all these are already in memory right?

Reply:
nothing final, will adapt all interfaces accordingly during refactoring driver and requests. not there yet